File: java/src/n03_race/Animal.java
package n03_race;

import java.util.Random;

public abstract class Animal implements Movable {
    public final String name;
    private final float minSpeed;
    private final float maxSpeed;
    private int stamina;

    public Animal(String name, float minSpeed, float maxSpeed, int stamina) {
        this.name = name;
        this.minSpeed = minSpeed;
        this.maxSpeed = maxSpeed;
        this.stamina = stamina;
    }

    public float move() {
        if (this.stamina == 0) {
            return this.minSpeed;
        } else {
            this.stamina--;
            return new Random().nextFloat(this.minSpeed, this.maxSpeed);
        }
    }

    public abstract void makeNoise();
}

File: dart/src/n03_race/animal.dart
import 'dart:math';

abstract class Animal {
  Animal(this.name, this._minSpeed, this._maxSpeed, this._stamina);

  final String name;
  final double _minSpeed;
  final double _maxSpeed;
  int _stamina;

  double move() {
    if (this._stamina == 0) {
      return this._minSpeed;
    } else {
      this._stamina--;
      return Random().nextDouble() * (this._maxSpeed - this._minSpeed) + this._minSpeed;
    }
  }

  void makeNoise();
}
